scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,1038 @@
+            for (Map.Entry<String, Serializable> entry : chosenChannel.entrySet()) {
+                currentKey = entry.getKey();
+                if (evaluateClosure(predicate, BiPredicate.class).test(currentKey, entry.getValue())) {
+
+                    chosenChannel.compute(currentKey,
+                                          (BiFunction<? super String, ? super Serializable, ? extends Serializable>) evaluateClosure(thenRemappingFunction,
+                                                                                                                                     BiFunction.class));
+                } else {
+                    chosenChannel.compute(currentKey,
+                                          (BiFunction<? super String, ? super Serializable, ? extends Serializable>) evaluateClosure(elseRemappingFunction,
+                                                                                                                                     BiFunction.class));
+
+                }
+            }
+
+            logWithContextAndPersist(taskid,
+                                     channel,
+                                     "Run conditionalReplaceAll, new content is: " + chosenChannel,
+                                     Level.DEBUG);
+
+        } catch (InvalidChannelException | CompilationException | IOException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new ClosureEvaluationException(EXCEPTION_WHEN_EVALUATING_CLOSURE + StackTraceUtil.getStackTrace(e));
+        }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean waitUntil(String originator, TaskId taskid, String channel, String key, String predicate)
+            throws InvalidChannelException, CompilationException {
+        Channel chosenChannel = getChannel(channel);
+        Serializable value = chosenChannel.get(key);
+        try {
+            return evaluateClosure(predicate, BiPredicate.class).test(key, value);
+        } catch (CompilationException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new ClosureEvaluationException(EXCEPTION_WHEN_EVALUATING_CLOSURE + StackTraceUtil.getStackTrace(e));
+        }
+    }
+
+    @Override
+    public boolean waitUntil(String originator, TaskId taskid, String channel, String key, String predicate,
+            long timeout) throws InvalidChannelException, CompilationException, TimeoutException {
+        // timeout is ignored inside the method implementation, it is handled by the runActivity method
+        return waitUntil(originator, taskid, channel, key, predicate);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public PredicateActionResult waitUntilThen(String originator, TaskId taskid, String channel, String key,
+            String predicate, String thenRemappingFunction)
+            throws InvalidChannelException, CompilationException, IOException {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)